network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_busy_duration_s",
+            "Histogram of busy time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Listener which converts published  updates from the OnChainConfig to ConnectivityRequests
+/// for the ConnectivityManager.
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+/// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a Validator Network.
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+
+    let mut updates = Vec::new();
+
+    // Collect the set of address updates.
+    let address_map = validator_keys
+        .iter()
+        .filter(|validator| {
+            match NetworkAddress::try_from(&validator.config().validator_network_address) {
+                Ok(_) => true,
+                Err(e) => {
+                    warn!(
+                        "Cannot parse validator network address from validator info: {}",
+                        e
+                    );
+                    false
+                }
+            }
+        })
+        .map(|validator| {
+            (
+                *validator.account_address(),
+                vec![
+                    NetworkAddress::try_from(&validator.config().validator_network_address)
+                        .expect("Extraction already passed"),
+                ],
+            )
+        })
+        .collect();
+
+    let update_address_req =
+        ConnectivityRequest::UpdateAddresses(DiscoverySource::OnChain, address_map);
+
+    updates.push(update_address_req);
+
+    // Collect the set of EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        validator_keys
+            .into_iter()
+            .map(|validator| {
+                (
+                    *validator.account_address(),
+                    NetworkPublicKeys {
+                        identity_public_key: validator.network_identity_public_key(),
+                        signing_public_key: validator.network_signing_public_key().clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+
+    updates
+}
+
+/// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a FullNode Network.
+fn extract_full_node_updates(full_node_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let full_node_infos = full_node_set.payload().to_vec();
+
+    let mut updates = Vec::new();
+
+    // Collect the set of updated addresses.
+    let address_map = full_node_infos
+        .iter()
+        .filter(|full_node| {
+            match NetworkAddress::try_from(&full_node.config().full_node_network_address) {
```

Comment:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map

filter_map will let you reduce code and run-time complexity (likely negligible on the latter) but most importanly remove this ugly expect statement below

note, I would probably do something like this:
```
map_filter(|fn|
  NetworkAddress::try_from(fn.config().full_node_network_address)
    .map_err(|e| {
       log_err;
       e
    })
    .ok()
)
```

or

```
let addr =  NetworkAddress::try_from(fn.config().full_node_network_address);
if let Err(err) = addr {
   log_err
}
addr.ok()
```

Reply:
oh of course david will stand up for this

Reply:
I'm used to use flat_map for this but anyway this two closures look ugly
```
flat_map(|node| {
  match NetworkAddress::try_from(&full_node.config().full_node_network_address) {
    Ok(addr) => Some((..)),
    Err(e) => {
       warn!(..);
       None
    }
  }
})
```

Reply:
I think my solution looks better than yours :P -- well the second one at least.

Reply:
flat_map wins.